common/persistence/nosql/nosql_execution_store_util.go

```diff
@@ -316,126 +316,50 @@ func (d *nosqlExecutionStore) prepareReplicationTasksForWorkflowTxn(domainID, wo
-	var tasks []*nosqlplugin.CrossClusterTask
-
-	for _, task := range crossClusterTasks {
-		var taskList string
-		var scheduleID int64
-		var targetCluster string
-		targetDomainID := domainID // default to source domain, can't be empty, since empty string is not valid UUID
-		targetDomainIDs := map[string]struct{}{}
-		var targetWorkflowID string
-		targetRunID := persistence.CrossClusterTaskDefaultTargetRunID
-		targetChildWorkflowOnly := false
-		recordVisibility := false
-
-		switch task.GetType() {
-		case persistence.CrossClusterTaskTypeStartChildExecution:
-			targetCluster = task.(*persistence.CrossClusterStartChildExecutionTask).TargetCluster
-			targetDomainID = task.(*persistence.CrossClusterStartChildExecutionTask).TargetDomainID
-			targetWorkflowID = task.(*persistence.CrossClusterStartChildExecutionTask).TargetWorkflowID
-			scheduleID = task.(*persistence.CrossClusterStartChildExecutionTask).InitiatedID
-
-		case persistence.CrossClusterTaskTypeCancelExecution:
-			targetCluster = task.(*persistence.CrossClusterCancelExecutionTask).TargetCluster
-			targetDomainID = task.(*persistence.CrossClusterCancelExecutionTask).TargetDomainID
-			targetWorkflowID = task.(*persistence.CrossClusterCancelExecutionTask).TargetWorkflowID
-			targetRunID = task.(*persistence.CrossClusterCancelExecutionTask).TargetRunID
-			if targetRunID == "" {
-				targetRunID = persistence.CrossClusterTaskDefaultTargetRunID
+func (d *nosqlExecutionStore) prepareNoSQLTasksForWorkflowTxn(
+	domainID, workflowID, runID string,
+	tasksByCategory map[persistence.HistoryTaskCategory][]persistence.Task,
+	tasksToAppend map[persistence.HistoryTaskCategory][]*nosqlplugin.HistoryMigrationTask,
+) (map[persistence.HistoryTaskCategory][]*nosqlplugin.HistoryMigrationTask, error) {
+	for c, tasks := range tasksByCategory {
+		switch c.ID() {
+		case persistence.HistoryTaskCategoryIDTransfer:
+			transferTasks, err := d.prepareTransferTasksForWorkflowTxn(domainID, workflowID, runID, tasks)
+			if err != nil {
+				return nil, err
 			}
-			targetChildWorkflowOnly = task.(*persistence.CrossClusterCancelExecutionTask).TargetChildWorkflowOnly
-			scheduleID = task.(*persistence.CrossClusterCancelExecutionTask).InitiatedID
-
-		case persistence.CrossClusterTaskTypeSignalExecution:
-			targetCluster = task.(*persistence.CrossClusterSignalExecutionTask).TargetCluster
-			targetDomainID = task.(*persistence.CrossClusterSignalExecutionTask).TargetDomainID
-			targetWorkflowID = task.(*persistence.CrossClusterSignalExecutionTask).TargetWorkflowID
-			targetRunID = task.(*persistence.CrossClusterSignalExecutionTask).TargetRunID
-			if targetRunID == "" {
-				targetRunID = persistence.CrossClusterTaskDefaultTargetRunID
+			for _, transfer := range transferTasks {
+				tasksToAppend[c] = append(tasksToAppend[c], &nosqlplugin.HistoryMigrationTask{
```

Comment:
I'd not rely on `tasksToAppend` being not nil. This function can create a new map, add items from `tasksToAppend` if exists, and populate the new map in these loops. This way it wouldn't alter the argument passed.

Reply:
The other way would be to rename `tasksToAppend` to `tasks`, populate it and don't return a map. Callers would expect this func to populate the passed in map. Right now it's a bit confusing